### ISSUE TITLE
Overlapping mob arenas can cause undefined behavior

### ISSUE THREAD

**Issue by Pugabyte** [garbagemule/MobArena#367]
Hi there.

One of our arenas is refusing to work. The config can be found [here](https://github.com/Pugabyte/ServerFiles/blob/master/MobArena/config.yml#L2034-L2173). 
Problem: It simply doesn't spawn the mobs. The waves progress but no mobs appear.
![image](https://cloud.githubusercontent.com/assets/12417961/24372899/2330cf82-12f5-11e7-893c-65c35e31787d.png)

The WorldGuard region is the same region that contains the other arenas, and the world is the same, so there shouldn't be any problems there.

I checked the spawnpoint coordinates, all are inside the arena.

Reloading displays no issues 
![image](https://cloud.githubusercontent.com/assets/12417961/24372724/7ea7cef2-12f4-11e7-85ea-07c56447ad84.png)
Nor are any errors displayed while the arena is running.

I have compared the config for that arena against a few others that do work, but I cannot see any difference that would affect the functionality of it. ([CourtOfClans](https://www.diffchecker.com/Qyga2IaH), [ProjectSpark](https://www.diffchecker.com/ohuYb1cY), [TrailsEnd](https://www.diffchecker.com/32Ecd2dt))

Versions:
![image](https://cloud.githubusercontent.com/assets/12417961/24373049/a94e45f4-12f5-11e7-8587-93ce96fda53f.png)

If you want to see it for yourself, the IP is play.bear-nation.net, version 1.12.2, and do `/ma join old-coliseum`

Thank you in advance for your help

**Comment by ct-martin** [comment]
@Pugabyte Hello, it's been a while since this issue was commented on, so if this issue is still relevant please comment and tag me (with the @) so I can repoen it. Thanks!

**Comment by Pugabyte** [comment]
@ct-martin This issue still appears to be a problem. Here is the IRC log of the conversation me and @garbagemule 
```
[14:13:43] <Pugabyte> when you have a chance slipcor and/or garbagemule https://github.com/garbagemule/MobArena/issues/367
[15:12:16] <garbagemule> Hey Pugabyte
[15:13:44] <garbagemule> This happens to people once in a while, and it has never been a MobArena bug. There are certain weird conditions that prevent mobs from spawning, and it's either the server configuration itself or a conflicting plugin. MobArena does what it can, but something else is removing its mobs or insta-killing them.
[15:14:58] <garbagemule> Historically, the problem has often been that people either by complete coincidence place the arena in the spawn region of the world, or somehow manage to move the spawn region to the arena during setup (I assume people sometimes type something like /setspawn or /spawn or something, expecting it to be a MobArena command).
[18:51:15] <Pugabyte> Okay, I will move the arena and see if that changes it
[12:19:38] <garbagemule> Pugabyte, cool cool! Let me know how it goes. I should try reproducing it at some point to see if MobArena can read any information off of the server or the world to determine this stuff before the arena starts. It'd be nice to just report on it immediately instead of people having to find out (or not find out) by happenstance.
[18:57:00] <Pugabyte> Still havent gotten around to moving it, but when I do I will make a copy instead of moving it, so if you ever want to send me a special build that has some debugging stuff I'd be happy to test it out on the arena
[17:57:25] <Pugabyte> Alright, we moved the mobarena and it works now. I still have the old one set up, still not working, so i'd be happy to test that for you
[17:57:40] <Pugabyte> I honestly have no idea what is in that area that's causing the mobs to not work
```

I took another look and it still refuses to work.

**Comment by ct-martin** [comment]
@Pugabyte thanks for confirming that this is still relevant (we were cleaning the issue tracker). I'm reopening the issue, but it sounds like @garbagemule has a rough idea of the cause. Could you post a list of plugins? It would help in trying to figure out what might be doing spawn protection or something.

@jflory7 do the docs have a known issue page? If so, ^

**Comment by Pugabyte** [comment]
I doubt this will be of much help... theres so many lol

`
PlugMan, AutomaticInventory, Murder, SimpleAPI, SidewaysLogFix, wget, AutoMessage, MinigameHandler, Chairs, ProtocolLib, CustomParticleEffects, bnCore, MVdWPlaceholderAPI, FastAsyncWorldEdit, PermissionsEx, FeatherBoard, WhatIs, Tab, ServerListPlus, PlaceholderAPI,
OneWayBlocks, InventoryChess, ActionBar, VoxelSniper, EditableSign, Alerts, TimeLock, StopPushing, dynmap, WorldEdit, OpenInv, WorldBorder, BuycraftX, Votifier, LibsDisguises, eZProtector, Super_Elytra, SQLibrary, DotCommand, Vault, TitleAdvanced, LWC, Quake, PlayerPoints, ChestCommands, PlayerVaults, MyCommand, GAListener, Herochat, Trading, PlotSquared, WorldGuard, ASkyBlock, Prism, Essentials, HeadDatabase, LiteBans, Dynmap-ASkyBlock, VanishNoPacket, Plot2Dynmap, CustomEnchantments, ConditionalPerms, SimplePets, Multiverse-Core, Citizens, mcMMO, MiniaturePets, MobArena, EssentialsProtect, Multiverse-Inventories, EssentialsAntiBuild, ParticleHats, EssentialsSpawn, WorldGuardExtraFlags, Denizen, PurpleIRC, PreciousStones, HeavySpleef, EnjinMinecraftPlugin, PermHelper, KillerMoney, HolographicDisplays, Depenizen, VoxelSniperRegions, Multiverse-NetherPortals, MultiWorldMoney, Minigames, Minigames-Regions
`

**Comment by ct-martin** [comment]
@Pugabyte can you check the spawn section of your essentials config and see if it's set? Also, the server.properties is another likely place.

**Comment by Pugabyte** [comment]
The server.properties spawn protection is off, and I know where it used to be, so that's not it. 
The `Essentials/spawn.yml` is in a different world and different coordinates.

**Comment by ct-martin** [comment]
@Pugabyte Thanks, I'll take a deeper look when I have a chance (off to bed soon).

**Comment by ct-martin** [comment]
@Pugabyte I got a chance to look through that list. My next suggestion to look at is the Multiverse world config and make sure the difficulty isn't set to peaceful

**Comment by Pugabyte** [comment]
@ct-martin All MobArenas are in the same world, so that wouldn't be it. But just to confirm, 
```
  gameworld:
    ==: MVWorld
    hidden: 'false'
    alias: gameworld
    color: WHITE
    style: NORMAL
    pvp: 'true'
    scale: '1.0'
    respawnWorld: ''
    allowWeather: 'true'
    difficulty: HARD
    spawning:
      ==: MVSpawnSettings
      animals:
        ==: MVSpawnSubSettings
        spawn: 'true'
        spawnrate: '-1'
        exceptions: []
      monsters:
        ==: MVSpawnSubSettings
        spawn: 'true'
        spawnrate: '-1'
        exceptions: []
    entryfee:
      ==: MVEntryFee
      amount: '0.0'
      currency: '-1'
    hunger: 'true'
    autoHeal: 'true'
    adjustSpawn: 'true'
    portalForm: ALL
    gameMode: SURVIVAL
    keepSpawnInMemory: 'true'
    spawnLocation:
      ==: MVSpawnLocation
      x: 117.0
      y: 7.0
      z: 415.0
      pitch: 0.0
      yaw: 0.0
    autoLoad: 'true'
    bedRespawn: 'true'
    worldBlacklist: []
    environment: NORMAL
    seed: '8182140439221091585'
    generator: 'null'
    playerLimit: '-1'
    allowFlight: 'true'
```

And while we're at it, since there's a spawn location in there, I made sure it's not near the arena:



```
p1: 547,0,330,0.0,0.0,gameworld
p2: 592,33,377,0.0,0.0,gameworld
```

**Comment by ct-martin** [comment]
@Pugabyte according to that Multiverse config, the spawn is `117, 7, 415`. Can you check if that's near your arena? Depending on the plugin you use to manage the spawn it sometimes uses a config option to change the spawnpoint instead of the actual world data, I think you might be falling victim to that.

**Comment by Pugabyte** [comment]
@ct-martin ![](https://i.imgur.com/lU10rLj.png) I already including that information in the previous comment

**Comment by ct-martin** [comment]
@Pugabyte same ;)
>  according to that Multiverse config, the spawn is 117, 7, 415

> Depending on the plugin you use to manage the spawn it sometimes uses a config option to change the spawnpoint instead of the actual world data, I think you might be falling victim to that.

**Comment by Pugabyte** [comment]
> Can you check if that's near your arena?

Like I said, I already provided that information.

**Comment by ct-martin** [comment]
@Pugabyte sorry, let me rephrase. There are two spawn points - one is a position that is coded into the actual data of the world (if you were to copy-paste the world into single player, that's where you would appear). The second kind is where a plugin has a config value that is used to make a virtual spawn point, similar to a warp, but is not the "true" spawn point of the world. Multiverse has the position of the world's spawn as `117, 7, 415`, which is different from the spawn points you provided. Additionally, you have two spawn points, whereas there can only be one "real" spawn point in a world. Could you check if the coordinate `117, 7, 415` is near your arena?

**Comment by Pugabyte** [comment]
![](https://i.imgur.com/lU10rLj.png)

I'm not sure how else to phrase this: I've already provided the location of my arena and made sure it's not near the multiverse spawn location. Additionally, I believe the coordinates are too far out to be default "singleplayer" spawn

**Comment by ct-martin** [comment]
After talking with @Pugabyte, coordinates in the image are for MobArena, not a spawn plugin. I updated the original post of the configs for clarity.

**Comment by ct-martin** [comment]
@Pugabyte My next best guess then is WorldGuard. Can you try making a WG region around your arena and setting the flag for spawning to true to test?

**Comment by Pugabyte** [comment]
![](https://i.imgur.com/aT2nGGv.png) No luck

**Comment by ct-martin** [comment]
I'm going to have to defer to @garbagemule then, sorry

**Comment by garbagemule** [comment]
Hey guys,

If the arena works in another location, then it has to be a result of world configuration or a conflicting plugin. It's usually WorldGuard, and even though one flag might be set correctly and everything , I've seen "internal conflicts" in WorldGuard regions where just the fact that two regions overlap is enough to mess things up.

I think the most pragmatic approach forward is to try removing plugins. @Pugabyte, can you make a local copy of your entire server configuration and just try removing _all_ plugins before booting up, then try and see if it has any effect? And by removing, I mean literally removing the jar-files from the plugins directory so they don't even load. If it works, it is a plugin conflict, and if it doesn't, it is a world configuration issue (which would be a good thing, since that means there is a chance that we can detect and report on it).

**Comment by Pugabyte** [comment]
This the exact process I went through:
- Copied my entire server to a new directory. 
- Removed all `.jar`s from `plugins` folder, except for MobArena.
- Edited `server.properties`, changed `port`, and changed `level-name` to the "gameworld" so that it would be the default world (since there is no Multiverse)
- Started and joined server
- Joined and started affected MobArena

No mobs spawned. 

- Quit MobArena and teleported to the location of it.
- Ran `/summon Zombie ~ ~ ~` - server was `Unable to summon object`
- Moved away from the area and ran the command again - Zombie spawned. 

![](https://i.imgur.com/dQJrGMw.png)

**Comment by Pugabyte** [comment]
So I continued to try to summon mobs, figuring out exactly where it stopped allowing me to spawn them. I found something very strange.

![](https://i.imgur.com/8Ee1BH2.png) 

The area perfectly outlines the arena, as though it was a region (remember, no plugins besides MobArena are installed). 

Even more interestingly, I compared this region to the positions stated in the configuration for the arena- *they match exactly*. 

```
p1: 547,0,330,0.0,0.0,gameworld
p2: 592,33,377,0.0,0.0,gameworld
```

The next logical step, of course, was to remove MobArena and restart the server. 

![](https://i.imgur.com/CS2xMR6.png)

The spawn was successful.

This is a MobArena issue

**Comment by Noahkiq** [comment]
could my test map be causing issues? my map seems to have the same regions

**Comment by Pugabyte** [comment]
welp. cant believe it was that simple. that's the issue.

**Comment by Noahkiq** [comment]
whoops.

i would keep this issue open though. maybe rename to "plugin not warning you if map regions are conflicting"

**Comment by garbagemule** [comment]
Thanks for the investigative work, guys.

I don't think MobArena's xenophobia is mentioned anywhere in this issue, but MobArena will do _everything in its power_ to prevent mobs from spawning in its regions unless it, itself, is the spawn reason (which it can actually never be truly certain of). It is _never_ going to allow you to spawn mobs in its region if the arena isn't running, so it is very important to test this stuff by actually running a session.

The spawn event is passed to all arenas. A given arena will always ignore spawn events outside of its region. Inside of its region, the event will be cancelled if it isn't a valid spawn reason (CUSTOM, BUILD_IRONGOLEM, BUILD_SNOWMAN). If the spawn reason is valid, the event will get "uncancelled" if the arena is running (attempting to override any cancellations from other plugins), or cancelled if the arena isn't running.

The TL;DR is that the order of arenas matters for reproducibility, and that arenas should always have disjoint regions and be separate physical structures. I don't think it's mentioned anywhere that regions can overlap, but it probably doesn't state anywhere that they can't either.

I see two different tasks spawn from this issue:

- The documentation should mention that overlapping arena regions and structure reuse is not supported, and that it _will_ result in undefined behavior.
- When the arenas are loaded from the config-file, the region of the arena currently being loaded must be checked for intersection with the regions of each loaded arena. In the case of an intersection, at the very least, a warning should be logged. Perhaps the conflicting arena should be disabled as well, just to really drive the point home. A similar message could be displayed in Setup Mode when defining new regions that overlap existing ones. This kind of check is extremely cheap with axis-aligned bounding boxes, so performance hits should be negligible.
